### ISSUE TITLE
Subclass of `tf.keras.Model` returns an empty list by `model.losses` 

### ISSUE THREAD

**Issue by siyuany** [tensorflow/tensorflow#37654]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow):  yes, as follows

```
import tensorflow as tf
from tensorflow import keras

class MyModel(keras.Model):
    def __init__(self, **kwargs):
        super(MyModel, self).__init__(**kwargs)
    def call(self, inputs):
        self.add_loss(tf.reduce_mean(tf.square(inputs)))
        return inputs

x = tf.placeholder(tf.float32, [None, 5])
model = MyModel()
model(x)

model.losses  # returns []
model._losses # returns [<tf.Tensor 'my_model/Mean:0' shape=() dtype=float32>]
```

- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): macOS Catalina, 10.15.3
- TensorFlow installed from (source or
binary): - TensorFlow version (use command below): 1.12.0
- Python version: Python 2.7.17 |Anaconda, Inc.| (default, Oct 21 2019, 14:10:59)
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin

**Describe the current behavior**

The current returned value of model.losses does not contain elements in model._losses

**Describe the expected behavior**

As we can see in the code of `Network.losses`, it should return 

```
list(set(relevant_conditional_losses + unconditional_losses + self._losses))
```

If it does not contain tensors in model._losses, `model.add_loss` will not take any effects.

**Comment by HopefulRational** [comment]
Hi. I am new here but want to contribute. 
So, @siyuany if this is what you are talking about : 
[https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/engine/network.py](url) line no : 400
(link not working... please copy paste in new tab)
then we see that in line 439, we are returning `unique_tensors + non_tensors`
So, will you please be little specific. Please dont mind if i am wrong coz i am totally new and just want to get a hang and contribute. Thanks.

**Comment by HopefulRational** [comment]
Hi @siyuany. Ok, are you saying we are not adding `self._losses` to the returned list?

**Comment by HopefulRational** [comment]
Hi @siyuany. I see that `self._losses` is initialized as empty list. Will you please tell where is `self._losses` getting updated ?
I mean in which file?

**Comment by siyuany** [comment]
> Hi. I am new here but want to contribute.
> So, @siyuany if this is what you are talking about :
> [https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/engine/network.py](url) line no : 400
> (link not working... please copy paste in new tab)
> then we see that in line 439, we are returning `unique_tensors + non_tensors`
> So, will you please be little specific. Please dont mind if i am wrong coz i am totally new and just want to get a hang and contribute. Thanks.

I think you're talking about the independent keras. But I refer to the keras in tensorflow, `tf.keras`. So the source codes are not the same.

**Comment by siyuany** [comment]
> Hi @siyuany. I see that `self._losses` is initialized as empty list. Will you please tell where is `self._losses` getting updated ?
> I mean in which file?

https://github.com/tensorflow/tensorflow/blob/5b900cfe4b3b848f577315a0dde09a729f770e95/tensorflow/python/keras/engine/network.py#L679

I think maybe here the property should return `list(set(losses + self._losses))`, or else losses added by `model.add_loss` will not be contained in the final list.

**Comment by HopefulRational** [comment]
Hi @siyuany. Thanks for such a quick reply.
I am sorry. I thought that the two keras are equivalent in implementation. I want to know where are we filling `model._losses`.
I think that in the `add_loss` : 
[https://github.com/tensorflow/tensorflow/blob/97103c3b9b447a9889be93fa36324b3ed3daf377/tensorflow/python/keras/engine/base_layer.py#L1172](url)
(again link is not working please copy paste link in new tab)
we are putting symbolic losses in `model._losses` and thats the only thing we are putting in `model._losses`. Am I right?

And would you please also tell me how to link a line like the way you linked line 679. Thanks.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please find the Gist [here](https://colab.research.google.com/gist/amahendrakar/77fc2e8c9039a0f310f847a0e9a5fef0/37654.ipynb). Thanks!

**Comment by siyuany** [comment]
It seems in tensorflow==1.15.0 the issue has been fixed.

**Comment by siyuany** [comment]
> Hi @siyuany. Thanks for such a quick reply.
> I am sorry. I thought that the two keras are equivalent in implementation. I want to know where are we filling `model._losses`.
> I think that in the `add_loss` :
> [https://github.com/tensorflow/tensorflow/blob/97103c3b9b447a9889be93fa36324b3ed3daf377/tensorflow/python/keras/engine/base_layer.py#L1172](url)
> (again link is not working please copy paste link in new tab)
> we are putting symbolic losses in `model._losses` and thats the only thing we are putting in `model._losses`. Am I right?
> 
> And would you please also tell me how to link a line like the way you linked line 679. Thanks.

Hi @HopefulRational

Yes, it's the same as I understand.

For the second question, You can just click the line no., click the three dots in the left, and select 'copy permalink'. And it seems you did in the same way.

**Comment by HopefulRational** [comment]
Hi @siyuany. Thanks for quick reply.
So, I see that `Network.losses` is not there.
Is it that in `base_layer.add_losses`, we are filling `_losses` and by `base_layer.losses` we are returning losses to solve the issue?

Thanks again. You have been very kind :-)

**Comment by HopefulRational** [comment]
@siyuany hi... could you please reply to my latest comment.. thanks :)

**Comment by siyuany** [comment]
> Thanks again. You have been very kind :-)

The bug had been fixed in higher version (tf>1.12), so I just closed this issue.
